### PR TITLE
docs: draft curriculum structure

### LINKED ISSUE #42938: Potential i18n bug when renaming curriculum files
A [recent PR of mine](https://github.com/freeCodeCamp/freeCodeCamp/pull/42885) renamed only the English version of a curriculum file. I was trying to `npm run develop` the site locally in Italian and got this:

Screen Shot 2021-07-19 at 9 35 11 PM

So my PR was in with the new filename in English and all the other locales had the old filename. So an error was thrown. I guess I'm wondering if PR's like that should rename all the languages. If so, we could perhaps add something to the action that says "Looks like you renamed an English file, you should rename all the others"

The other languages will be updated on the next pull from crowdin I suppose, fixing everything. But this would block us from deploying until that is fixed, no? So it seems like we could try and catch that. Not sure, maybe I'm missing something.

### PR DESCRIPTION
**Pull request by naomi-lgbt**
Checklist:

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

Closes #42938 I think?

Starting the documentation for how to rename things in the curriculum.

@ojeytonwilliams let's touch base on a couple of things? I need to get a better understanding of how `dasherize` is used vs. the `dashedName` frontmatter, and where the superblock data comes from.

### CODE REVIEW

**Review by ojeytonwilliams — COMMENTED** [review]
I think the `intro.json`s need updating too.

File: docs/curriculum-file-structure.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+When renaming a superblock, you need to:
+
+1. Rename the superblock folder in the `english` directory.
+2. Rename the superblock folder in *all* language directories.
+3. For each block within that superblock, update the `superBlock` value in the `meta.json` file.
+4. Rename the superblock folder in `client/src/pages/learn`.
+5. Update the `index.md` file in the above folder, changing the `superBlock` value to match the new name.
+6. For each block folder within the above, update the `index.md` to use the correct `superBlock` value.
+7. Update `client/src/redux/index.ts` to use the correct `title` and `certSlug` for the superblock - the `certSlug` being the new folder name.
+8. Update `client/src/resources/cert-and-project.map.ts` in the same manner, updating the path for the cert at the top of the file, and the `title` and `certSlug` for the certificate and the nested `projects` values.
```
Comment (ShaunSHamilton) [inline docs/curriculum-file-structure.md:51]: We should stress there are very rare instances where renaming the `certSlug` is desired - it is a last resort.

Comment (ojeytonwilliams) [reply docs/curriculum-file-structure.md:51]: 100%. As we discussed in the back end development issue, each time we do it it adds to the maintenance overhead.

Comment (nhcarrigan) [reply docs/curriculum-file-structure.md:51]: Does it not cause an issue if we rename a superblock without changing the `certSlug`??

File: docs/curriculum-file-structure.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+3. Change the name of the block folder in the `_meta` directory. 
+4. Update the `dashedName` property for that block's `meta.json` file.
+
+### Renaming a Challenge
+
+When renaming a single challenge file, you need to:
+
+1. Change the name of the challenge file in the `english` directory.
+2. Change the name of the challenge file in *all* of the other language directories to match.
+3. Change the value of the `dashedName` property in that challenge's frontmatter.
```
Comment (moT01) [inline docs/curriculum-file-structure.md:73]: And the `title` of the English version

File: docs/curriculum-file-structure.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+4. Optionally, update the name of the challenge in the relevant `meta.json` file. The challenge names here are not used in the build, but provide a user-friendly way to identify the challenge order.
```
Comment (moT01) [inline docs/curriculum-file-structure.md:74]: Yea, it seems like you were going for instructions for just renaming the files - and not particularly renaming everything needed to change how it's displayed. I feel like we should be renaming everything - or we may end up with a challenge with one name, and another name in the meta, and confusion sets in.

```
Step 5: if it's a certificate project rename, update the yaml
file in `curriculum/english/12-certificates/<superBlock>`
to the new name
```

**Review by moT01 — COMMENTED** [review]
It's looking good @nhcarrigan 👍 I hope this review isn't overkill. 

I tried to follow these steps to rename things. I left some comments about it, but it seems like these are more instructions to rename the files and variables within the codebase to make sure things work. But don't get all the values to change what's displayed. I think those should be included. I left comments on some of the things that are missing - it may not be everything.

If we were to merge this PR as-is I think it would be helpful, as it points you to most of the spots the files are located - which would be helpful - and currently, we don't have anything else. I would be willing to make a commit with some of my suggestions, if you want - let me know.

**Comment by nhcarrigan** [comment]
Definitely open to more clarity, so yes please feel free to push commits.

This is not something we need to do very often, so having clear step by step guidelines will help prevent a lot of future headache I think.

**Comment by moT01** [comment]
I think I'll try and do that right now while it's fresh on my mind.

File: docs/curriculum-file-structure.md (hunk 1 above)
Comment (ShaunSHamilton) [reply docs/curriculum-file-structure.md:51]: > Does it not cause an issue if we rename a superblock without changing the `certSlug`??

Yes and no. The issue with renaming the `certSlug` is we need redirects for the certificates.

Comment (moT01) [reply docs/curriculum-file-structure.md]: Should be resolved with the commit I added

**Comment by moT01** [comment]
I may have went a little overboard, but the instructions are much more granular now. I separated the renaming of a cert and renaming a superblock.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I'm not sure the best place to put it, but we should add a warning about redirects.  I.e. that any time we change a slug, we need to add redirects from the old to the new.

**Review by ojeytonwilliams — DISMISSED** [review]
I think this is in good shape.  If there are any errors, it should become obvious when someone tries to use the docs and we can fix if and when we discover them.

File: docs/_sidebar.md [hunk 4]
```diff
@@ -19,6 +19,7 @@
   - [Work on localized client web app](how-to-work-on-localized-client-webapp.md)
   - [Catch outgoing emails locally](how-to-catch-outgoing-emails-locally.md)
   - [Test translations locally](how-to-test-translations-locally.md)
+  - [Understand the Curriculum File Structure](curriculum-file-structure.md)
```
Comment (ShaunSHamilton) [inline docs/_sidebar.md:22]:
```suggestion
  - [Understand the curriculum file structure](curriculum-file-structure.md)
```

File: docs/curriculum-file-structure.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+├─ _meta/
+│  ├─ {block}/
+│  │  ├─ meta.json
+├─ {language}/
+│  ├─ {superBlock}/
+│  │  ├─ {block}/
+│  │  │  ├─ {challenge}.md
+```
+
+## The `_meta` directory
```
Comment (ShaunSHamilton) [inline docs/curriculum-file-structure.md:30]:
```suggestion
## The `_meta` Directory
```

File: docs/curriculum-file-structure.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+1. Rename the `curriculum/challenges/_meta/{superBlock}-certficate` folder to the new name
+1. In the `meta.json` file of that folder, rename the values in `name`, `dashedName`, and `challengeOrder` to the new cert name
+1. In `curriculum/challenges/english/12-certificate`, rename the `{superBlock}-certificate` folder, and the YAML file within it, to the new name.
+1. In the YAML file, change the `title` to the new name
+1. Rename the file and folder from step 3 for the rest curriculum languages
+1. Update `client/src/redux/index.ts` to use the correct `title`
+1. Optionally, update the `certSlug` for the superblock in the same file. **Note** that renaming a `certSlug` will change the URL for certifications and should only be done with careful consideration.
+1. Update the `title` in `client/src/resources/cert-and-project.map.ts` to the new value. **Note** that changing the `title` here **will break** the superBlock page for the associated certification. It relies on the superBlock title to match the certification title. You will likely want to rename the superBlock at the same time
```
Comment (ShaunSHamilton) [inline docs/curriculum-file-structure.md:52]:
```suggestion
1. Update the `title` in `client/src/resources/cert-and-project-map.ts` to the new value. **Note** that changing the `title` here **will break** the superBlock page for the associated certification. It relies on the superBlock title to match the certification title. You will likely want to rename the superBlock at the same time
```

File: docs/curriculum-file-structure.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+1. Rename the superBlock folder in the `curriculum/challenges/english` directory.
+1. Rename the superBlock folder in *all* other `curriculum/challenges/{language}` directories.
+1. For each block within that superBlock, update the `superBlock` value in the `meta.json` file to its dashedName. You don't need to rename any folders here. Do that when renaming a block
+1. Rename the superblock folder in `client/src/pages/learn`.
+1. Update the `index.md` file in the above folder, changing the `title` and `superBlock` values to the new name.
+1. For each block folder within the above, update the `index.md` to use the correct `superBlock` value.
+1. In the `client/src/resources/cert-and-project.map.ts` file, update the path for the cert at the top of the file, and the `title` value for that superBlock. **Note** that changing the `title` here **will break** the ability to view the actual certification for this superBlock. It relies on the superBlock title to match the certification title. You will likely want to rename the certification at the same time
```
Comment (ShaunSHamilton) [inline docs/curriculum-file-structure.md:72]:
```suggestion
1. In the `client/src/resources/cert-and-project-map.ts` file, update the path for the cert at the top of the file, and the `title` value for that superBlock. **Note** that changing the `title` here **will break** the ability to view the actual certification for this superBlock. It relies on the superBlock title to match the certification title. You will likely want to rename the certification at the same time
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Minor changes. Just to mention, if anyone is feeling particularly rambunctious to make the change, some bullet points end in full stops, others do not.

**Review by ShaunSHamilton — APPROVED** [review]
Thanks, for feeling rambunctious, Nich 👍

**Review by ojeytonwilliams — APPROVED** [review]

**Pull request merged**